src/plugins_3rdparty/primer3/src/Primer3Plugin.cpp

```diff
@@ -92,62 +97,36 @@ void Primer3ADVContext::initViewContext(GObjectView* v) {
-    assert(av);
+    SAFE_POINT(av != nullptr, L10N::nullPointerError("AnnotatedDNAView"), );
 
     ADVSequenceObjectContext* seqCtx = av->getActiveSequenceContext();
-    assert(seqCtx->getAlphabet()->isNucleic());
-    {
-        Primer3TaskSettings defaultSettings;
-        {
-            QList<U2Region> sizeRange;
-            sizeRange.append(U2Region(150, 101));  // 150-250
-            sizeRange.append(U2Region(100, 201));  // 100-300
-            sizeRange.append(U2Region(301, 100));  // 301-400
-            sizeRange.append(U2Region(401, 100));  // 401-500
-            sizeRange.append(U2Region(501, 100));  // 501-600
-            sizeRange.append(U2Region(601, 100));  // 601-700
-            sizeRange.append(U2Region(701, 150));  // 701-850
-            sizeRange.append(U2Region(851, 150));  // 851-1000
-            defaultSettings.setProductSizeRange(sizeRange);
-        }
-        defaultSettings.setDoubleProperty("PRIMER_MAX_END_STABILITY", 9.0);
-        defaultSettings.setAlignProperty("PRIMER_MAX_TEMPLATE_MISPRIMING", 1200);
-        defaultSettings.setAlignProperty("PRIMER_PAIR_MAX_TEMPLATE_MISPRIMING", 2400);
-        defaultSettings.setIntProperty("PRIMER_LIBERAL_BASE", 1);
-        defaultSettings.setDoubleProperty("PRIMER_WT_POS_PENALTY", 0.0);
-        defaultSettings.setIntProperty("PRIMER_FIRST_BASE_INDEX", 1);
-
-        QObjectScopedPointer<Primer3Dialog> dialog = new Primer3Dialog(defaultSettings, seqCtx);
-        dialog->exec();
-        CHECK(!dialog.isNull(), );
-
-        if (QDialog::Accepted == dialog->result()) {
-            Primer3TaskSettings settings = dialog->getSettings();
-            U2OpStatusImpl os;
-            QByteArray seqData = seqCtx->getSequenceObject()->getWholeSequenceData(os);
-            CHECK_OP_EXT(os, QMessageBox::critical(QApplication::activeWindow(), L10N::errorTitle(), os.getError()), );
-            settings.setSequence(seqData,
-                                 seqCtx->getSequenceObject()->isCircular());
-            QString err = dialog->checkModel();
-            if (!err.isEmpty()) {
-                QMessageBox::warning(QApplication::activeWindow(), dialog->windowTitle(), err);
-                return;
-            }
-            bool objectPrepared = dialog->prepareAnnotationObject();
-            if (!objectPrepared) {
-                QMessageBox::warning(QApplication::activeWindow(), tr("Error"), tr("Cannot create an annotation object. Please check settings"));
-                return;
-            }
-            const CreateAnnotationModel& model = dialog->getCreateAnnotationModel();
-            AppContext::getTaskScheduler()->registerTopLevelTask(new Primer3ToAnnotationsTask(settings, seqCtx->getSequenceObject(), model.getAnnotationObject(), model.groupName, model.data->name, model.description));
-        }
+    SAFE_POINT(seqCtx != nullptr, L10N::nullPointerError("ADVSequenceObjectContext"), );
+    
+    Primer3Dialog dialog(seqCtx);
```

Comment:
Do not use on-stack dialogs, use QSharedPointer pattern so it will be safe to remove the dialog with a parent

Reply:
I can do it, but I don't get, why is that needed? Now we have the best possible situation - `QDialog` will be removed itself from than the function pops from the call stack (because it has a lifetime only in `{ .. }` it was defined in). By the way, `QSharedPointer` does not relates to the parent lifetime in our case - it well be deleted at the same time as now, when the pointer itself dies.